gapfill/gapfill.py

```diff
@@ -1,22 +1,227 @@
+                gap_df.loc[observed_rows, 'FCH4']
+            )
+        # TODO: Should we keep uncertainty around observed values?
+        gap_df.loc[observed_rows, f'FCH4_F_UNCERTAINTY'] = 0
+
+        outpath = gap_dir / f"{model}_{predictor_subset}_{distribution}.csv"
+        print(f"Writing gapfilled data to {outpath}")
+        gap_df.to_csv(outpath, index=False)
+
+        # For budget estimates, convert from nmol m-2 s-1 to g C m-2 halfhour-1
+        for fch4_column in fch4_columns + ['FCH4', 'FCH4_F_UNCERTAINTY']:
+            gap_df[fch4_column] = gap_df[fch4_column].apply(
+                lambda fch4: fch4*60*30*12.0107*10**-9
+            )
+
+        # Get date ranges to compute annual budget
+        if site in budget_date_ranges:
+            site_budget_date_ranges = [
+                [
+                    dt.datetime(
+                        int(range_start.split('-')[1]),
+                        int(range_start.split('-')[0]),
+                        1
+                    ),
+                    dt.datetime(
+                        int(range_end.split('-')[1]),
+                        int(range_end.split('-')[0]),
+                        1
+                    ) + relativedelta(months=1)
+                ]
+                for [range_start, range_end] in budget_date_ranges[site]
+            ]
+        else:
+            annual_sum_year_start = gap_df['Year'].min()
+            annual_sum_year_end = gap_df['Year'].max()
+            site_budget_date_ranges = [
+                [
+                    dt.datetime(year, 1, 1),
+                    dt.datetime(year+1, 1, 1)
+                ]
+                for year in range(
+                    annual_sum_year_start,
+                    annual_sum_year_end + 1
+                )
+            ]
+
+        # Compute annual budgets between date ranges
+        gap_df["TIMESTAMP_END"] = pd.to_datetime(
+            gap_df["TIMESTAMP_END"],
+            format='%Y%m%d%H%M'
+        )
+        budget_dict = defaultdict(list)
+        for [range_start, range_end] in site_budget_date_ranges:
+            annual_gap_df = gap_df[
+                (gap_df['TIMESTAMP_END'] >= range_start) &
+                (gap_df['TIMESTAMP_END'] < range_end)
+            ]
+            annual_budgets = annual_gap_df[fch4_columns].sum()
+
+            budget_mean = annual_budgets.mean()
+            budget_uncertainty = annual_budgets.std() * 1.96
+            print(
+                f"Budget between {range_start.date()} and {range_end.date()}" +
+                f" is {budget_mean:.2f} +/- {budget_uncertainty:.2f}"
+            )
+
+            budget_dict['range_start'].append(range_start)
+            budget_dict['range_end'].append(range_end)
+            budget_dict['budget_mean'].append(budget_mean)
+            budget_dict['budget_uncertainty'].append(budget_uncertainty)
+
+        # TODO: How did we combine uncertainties per year? average?
```

Comment:
@gavinmcnicol  Right now the code computes the budget with uncertainty per year. How did we combine them to get the average annual budgets in the table?

Reply:
I computed the multi-year mean.

Reply:
Got it! Will take the mean over the per-year budgets and uncertainties then.